cdc/scheduler/balancer.go

```diff
@@ -0,0 +1,142 @@
+// Copyright 2021 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package scheduler
+
+import (
+	"math"
+
+	"github.com/pingcap/ticdc/cdc/model"
+	"github.com/pingcap/ticdc/cdc/scheduler/util"
+	"go.uber.org/zap"
+)
+
+// balancer is used to support the rebalance process, in which
+// some victims are chosen and de-scheduled. Later, the victims
+// will be automatically rescheduled, during which the target captures
+// will be chosen so that the workload is the most balanced.
+type balancer interface {
+	// FindVictims returns a set of possible victim tables.
+	// Removing these tables will make the workload more balanced.
+	FindVictims(
+		tables *util.TableSet,
+		captures map[model.CaptureID]*model.CaptureInfo,
+	) (tablesToRemove []model.TableID)
+
+	// FindTarget returns a target capture to add a table to.
+	FindTarget(
+		tables *util.TableSet,
+		captures map[model.CaptureID]*model.CaptureInfo,
+	) (minLoadCapture model.CaptureID, ok bool)
+}
+
+// tableNumberBalancer implements a balance strategy based on the
```

Comment:
Is it feasible to add a fine-grained balancer implementation? For example, we can rebalance tables based on their load such as the total number of events processed over a period of time

Reply:
This design is not incompatible with a fine-grained design. If we can make the processors report their current workload more accurately, we will be able to find victims based on their actual workload.